src/main/java/com/google/sps/DataServlet.java

```diff
@@ -179,118 +193,162 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     } else {
-      // If queried and queried next contain the same nodes, then no reason to regenerate the graph
+      // If queried and queriedNext contain the same nodes, then there is no reason
+      // to regenerate the graph
       truncatedGraphNext =
           queried.equals(queriedNext)
               ? truncatedGraph
-              : currDataGraph.getReachableNodes(queriedNext, depthNumber);
+              : currDataGraph.getReachableNodes(queriedNext, radius);
     }
 
     // If we are not filtering the graph or limiting its depth, show all mutations of all nodes
     if (nodeNames.size() == 0
-        && tokenParam.length() == 0
+        && tokenNameParam.length() == 0
         && truncatedGraph.equals(currDataGraph.graph())) {
       filteredMutationIndices = defaultIndices;
       filteredDiff = diff;
     } else {
-      // Get the names of all the displayed nodes and find all indices of mutations
-      // that mutate any of them
       Set<String> truncatedGraphNodeNames = getNodeNamesInGraph(truncatedGraph);
       Set<String> truncatedGraphNodeNamesNext = getNodeNamesInGraph(truncatedGraphNext);
 
       // A set containing a indices where nodes currently displayed on the graph
       // or queried are mutated
       Set<Integer> mutationIndicesSet = new HashSet<>();
 
+      // Add all mutations relevant to on-screen nodes
       mutationIndicesSet.addAll(
           findRelevantMutations(truncatedGraphNodeNamesNext, mutationIndicesMap, mutList));
-      mutationIndicesSet.addAll(getMutationIndicesOfToken(tokenParam, mutList));
+
+      // Add all mutations relevant to the queried token, computing and caching it if it
+      // hasn't been done already
+      if (!tokenIndicesMap.containsKey(tokenNameParam)) {
+        tokenIndicesMap.put(tokenNameParam, getMutationIndicesOfToken(tokenNameParam, mutList));
+      }
+      mutationIndicesSet.addAll(tokenIndicesMap.get(tokenNameParam));
+
+      // Add all mutations relevant to the queried node names
       mutationIndicesSet.addAll(findRelevantMutations(nodeNames, mutationIndicesMap, mutList));
       filteredMutationIndices = new ArrayList<>(mutationIndicesSet);
       Collections.sort(filteredMutationIndices);
 
-      // Show mutations relevant to nodes that used to have the token but
-      // might not exist anymore and the queried nodes
+      // Show mutations relevant to nodes that are related to on-screen nodes, nodes that
+      // used to/still have the queried token and any queried nodes
       filteredDiff = filterMultiMutationByNodes(diff, Sets.union(truncatedGraphNodeNames, queried));
     }
+
+    /*
+     ***********************
+     * Error Handling
+     ***********************
+     */
+
     // We set the headers in the following 4 scenarios:
-    // The searched node is not in the graph and is never mutated
+    // truncatedGraph.nodes().size() == 0 means something was queried but wasn't found
```

Comment:
nit: I don't think you need to copy the actual code here. You can inline the comment with the code:
```
if (truncatedGraph.nodes().size() == 0 // something was queried but wasn't found in the graph
    && filteredMutationIndices.size() == 0) { // the searched object is never mutated
```